components/supervisor/pkg/supervisor/top.go

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package supervisor
+
+import (
+	"context"
+	"io/ioutil"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	linuxproc "github.com/c9s/goprocinfo/linux"
+	"github.com/gitpod-io/gitpod/supervisor/api"
+)
+
+// Top provides workspace resources status information.
+func Top(ctx context.Context) (*api.ResourcesStatusResponse, error) {
+	memory, err := resolveMemoryStatus()
+	if err != nil {
+		return nil, err
+	}
+	cpu, err := resolveCPUStatus()
+	if err != nil {
+		return nil, err
+	}
+	return &api.ResourcesStatusResponse{
+		Memory: memory,
+		Cpu:    cpu,
+	}, nil
+}
```

Comment:
We should harmonise this with the way ws-daemon does it.
Either copy over some of that code, or stick it in something like `common-go/cgroup`.

@Furisto @utam0k could you guide here how to proceed?

Reply:
At the moment the code in ws-daemon is not written to be used by other components. Factoring this out would be equivalent to starting our own cgroup library. Considering that the other options like using containerd's cgroup are awkward to use I still think this is the best way forward. This does not need to be done in this PR though.

Reply:
Could you point me to relevant code please? 🙏 

There also can be slight difference, i.e. I subtract evict-able memory to give a user more usefull data, i.e. simlar to what GCP showing. I am not sure what ws-daemon does.

Reply:
CPU: 
v1; https://github.com/gitpod-io/gitpod/blob/main/components/ws-daemon/pkg/cpulimit/cfs.go
v2: https://github.com/gitpod-io/gitpod/blob/main/components/ws-daemon/pkg/cpulimit/cfs_v2.go

Memory: 
v1: https://github.com/gitpod-io/gitpod/blob/main/components/ws-daemon/pkg/cgroup/plugin_cachereclaim.go
v2: n.a.

IO:
v1: https://github.com/gitpod-io/gitpod/blob/main/components/ws-daemon/pkg/cgroup/plugin_iolimit_v1.go
v2: https://github.com/gitpod-io/gitpod/blob/main/components/ws-daemon/pkg/cgroup/plugin_iolimit_v2.go

Reply:
> We should harmonise this with the way ws-daemon does it.

After looking at the code I don't think it can be harmonised without refactoring, ws daemon does not have some functions like getting used memory without evictable caches. Instead it compare page caches after trying to reclaim memory. It also has special cases for different state of workspaces which is not applicable inside a workspace.

I'm not sure refactoring will make such sense trying to create a reusable library right now. Code in ws-daemon is not really source of truth, i.e. there should not be a situation when you change one and need to change another.

Reply:
I filed https://github.com/gitpod-io/gitpod/issues/10043 to look how to refactor it nicely later.

Reply:
@akosyakov Great PR. I have a question do you have a specific plan to support cgroup v2? As far as I know, In the case of cgroup v2, the supervisor may don't see the cgroup value of the pod.
https://github.com/gitpod-io/gitpod/blob/cf2c0aafa9d38973e3b599637e8c7256d8d3701b/components/ws-daemon/pkg/iws/iws.go#L915-L919
I recommend checking to be sure. You can build the gcp image with cgroup v2 by changing this line. And you can use a gcp image you build on workspace-preview.
https://github.com/gitpod-io/gitpod-packer-gcp-image/blob/ef5be5d24542142f39fc2aa6748761d2f2177124/setup.sh#L111-L112

Reply:
You can use `gitpod-k3s-202205162340` I built as a cgroup v2 version image.
on workspace
```console
root@gitpodio-templategolang-twmrukrpk05:/sys/fs/cgroup# cat /sys/fs/cgroup/memory.max 
max
root@gitpodio-templategolang-twmrukrpk05:/sys/fs/cgroup# cat /sys/fs/cgroup/memory.high
max
```
I make sure that the workspace's pod was limited
```console
    resources:
      limits:
        cpu: "2"
        memory: 4Gi
      requests:
        cpu: "1"
        memory: 2Gi
```

Reply:
So, I guess what I am saying is that in cgroup v2 the memory limit is not visible from the supervisor. This could be an issue with cgroup v2 as the JVM probably can't see it either.

Reply:
@utam0k Is it always like that or it is how we implement it?

I asked already on other channels whether instead of relying on files we can actually ask container runtime endpoints to collect metrics and provide results via workspace kits somehow?

I wonder also why ws-daemon does not use container runtime endpoints but need to read from the disk?

Reply:
Anyway we discussed with @iQQBot to land this PR as is today and then tomorrow start working on https://github.com/gitpod-io/gitpod/issues/10043 and support of cgroup v2 in supervisor. So I think we will need to research it as a part of it.

Reply:
> I wonder also why ws-daemon does not use container runtime endpoints but need to read from the disk?

Thanks for your feedback! Sorry, I couldn't get what `the disk` is.

Reply:
> Thanks for your feedback! Sorry, I couldn't get what the disk is.

I mean container runtime has [such API](https://github.com/opencontainers/runc/blob/067aaf8548d78269dcb2c13b856775e27c410f9c/libcontainer/container.go#L94-L95) which can return usage/limit stats for cpu and memory: https://github.com/opencontainers/runc/blob/067aaf8548d78269dcb2c13b856775e27c410f9c/libcontainer/cgroups/stats.go#L158-L161

Supervisor cannot reach this endpoint, so we have to read cgroup from filesystem. I am asking whether ws-daemon can rely on it instead and provide this information to supervisor somehow?

Reply:
The API you have linked would mean to execute runc stat and then parse the output, this is not a service endpoint. We also need to write to the cgroup filesystem not only read from it and we should strive to do so in a runtime agnostic manner. We could ask this information from containerd but if we allow the option to use any CRI compatible runtime in the future (like CRI-O) this would require us to adapt the code again.

Reply:
ok, sorry i don't really know it and though opencontainers runc is an abstraction over different runtimes and using this API oppositely will make things more stable.

Reply:
@csweichel Would it be alright to continue as suggested [here](https://github.com/gitpod-io/gitpod/pull/9796#discussion_r874502070)? Could you approve it please if so? 🙏